be/src/formats/csv/array_reader.cpp

```diff
@@ -0,0 +1,109 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "array_reader.h"
+
+#include "column/array_column.h"
+
+namespace starrocks::vectorized::csv {
+
+bool StandardArrayReader::read_string(const std::unique_ptr<Converter>& elem_converter, Column* column, Slice s, const Converter::Options& options) const {
+    if (s.size < 2) {
+        return false;
+    }
+    if (s[0] != '[' || s[s.size - 1] != ']') {
+        return false;
+    }
+    s.remove_prefix(1);
+    s.remove_suffix(1);
+
+    auto* array = down_cast<ArrayColumn*>(column);
+    auto* offsets = array->offsets_column().get();
+    auto* elements = array->elements_column().get();
+
+    std::vector<Slice> fields;
+    if (!s.empty() && !split_array_elements(s, &fields)) {
+        return false;
+    }
+    size_t old_size = elements->size();
+    Converter::Options sub_options = options;
+    sub_options.invalid_field_as_null = false;
+    DCHECK_EQ(old_size, offsets->get_data().back());
+    for (const auto& f : fields) {
+        if (!elem_converter->read_quoted_string(elements, f, sub_options)) {
+            elements->resize(old_size);
+            return false;
+        }
+    }
+    offsets->append(old_size + fields.size());
+    return true;
+}
+
+bool StandardArrayReader::split_array_elements(Slice s, std::vector<Slice>* elements) const {
+    bool in_quote = false;
+    int array_nest_level = 0;
+    elements->push_back(s);
+    for (size_t i = 0; i < s.size; i++) {
+        char c = s[i];
+        // TODO(zhuming): handle escaped double quotes
+        if (c == '"') {
+            in_quote = !in_quote;
+        } else if (!in_quote && c == '[') {
+            array_nest_level++;
+        } else if (!in_quote && c == ']') {
+            array_nest_level--;
+        } else if (!in_quote && array_nest_level == 0 && c == _array_delimiter) {
+            elements->back().remove_suffix(s.size - i);
+            elements->push_back(Slice(s.data + i + 1, s.size - i - 1));
+        }
+    }
+    if (array_nest_level != 0 || in_quote) {
+        return false;
+    }
+    return true;
+}
+
+bool HiveTextArrayReader::read_string(const std::unique_ptr<Converter>& elem_converter, Column* column, Slice s, const Converter::Options& options) const {
+    auto* array = down_cast<ArrayColumn*>(column);
```

Comment:
almost same logic to StandardArrayrReader?

Reply:
maybe this piece of code is common:
1. split array elements into vector of slices
2. use elemnt_converted to parse slice

Reply:
The array format of Hive is very different from the standard array format. So writing `read_string` separately here will be easier to maintain in the future.

Reply:
Did you notice that in `StandardArrayReader` is use `read_quoted_string()`, and in `HiveTextArrayReader` is use `read_string`.

Reply:
OK. For one reader,  is only one of them `read_string` or `read_quoted_string` be used? If only one is used, possbily we could unify it ?

Reply:
I had to rewrite a new version. PTAL.